src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala

```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
     val blocks = ListBuffer[PhaseBlock]()
     logger.info(s"Found ${phaseBlocks.size} phase block")
-    phaseBlocks.values.toList.sortBy { b => (b.start, -b.length) } match {
-      case Nil                    => Unit
-      case block :: Nil           => blocks.append(block)
-      case block :: currentBlocks =>
-        var left = block
-        currentBlocks.foreach { right =>
-          require(left.getStart <= right.getStart)
-          if (left.overlaps(right)) {
-            if (!modifyBlocks) {
-              throw new IllegalStateException(s"Block $left overlaps $right")
-            }
-            if (left.encloses(right)) { // keep the left block only
-              logger.info(s"Removing $right enclosed in $left")
-              // do nothing
-            }
-            else if (right.encloses(left)) { // keep the right block
-              logger.info(s"Removing $left enclosed in $right")
-              left = right
-            }
-            else if (left.length < right.length) { // truncate left
-              logger.info(s"Truncating $left which overlaps $right");
-              blocks.append(left.copy(end=right.getStart-1))
-              left = right
-            }
-            else { // truncate right
-              logger.info(s"Truncating $right which overlaps $left");
-              blocks.append(left)
-              left = right.copy(start=left.getEnd+1)
-            }
-          }
-          else {
-            blocks.append(left)
-            left = right
-          }
+
+    var left = blockSet.pollFirst()
+    while (!blockSet.isEmpty) {
+      val right = blockSet.pollFirst()
+
+      if (right.getStart < left.getStart) {
+        throw new IllegalArgumentException(s"left: $left right: $right")
+      }
+      require(left.getStart <= right.getStart)
+      if (left.overlaps(right)) {
+        if (!modifyBlocks) {
+          throw new IllegalStateException(s"Block $left overlaps $right")
+        }
+        if (left.encloses(right)) { // keep the left block only
+          logger.info(s"Removing $right enclosed in $left")
+          // do nothing
+        }
+        else if (right.encloses(left)) { // keep the right block
+          logger.info(s"Removing $left enclosed in $right")
+          left = right
+        }
+        else if (left.length < right.length) { // truncate left
+          logger.info(s"Truncating $left which overlaps $right");
+          blocks.append(left.copy(end=right.getStart-1))
+          left = right
         }
-        // don't forget the last one
+        else { // truncate right
+          logger.info(s"Truncating $right which overlaps $left");
+          blocks.append(left)
+          // Since the start position of the right block is changed, and we are going to use it in the next iteration,
+          // then re-insert it into the set and poll.
+          blockSet.add(right.copy(start=left.getEnd+1))
+          left = blockSet.pollFirst()
```

Comment:
This is the part that was buggy.  We require that `left.start <= right.start` at the start of the loop. If we assign `left = right`, then in the requirement many not hold.  An example, which I will add to a test, would be: `(1, 10), (8, 12), (9, 13)`.  If we do as you suggested, then in the first iteration we would truncate the right to `(11, 12)` and assign `left = (11,12)`, and in the next loop we would assign `right = (9, 13)` (`right = blockSet.pollFirst()`), and now our requirement that `left.start <= right.start` at the start of the loop fails.  What we do instead is re-insert `right = (11, 12)` into the tree-set, then call `pollFirst` which assigns `left = (9, 13)`, and so in the next loop, `right = (11, 12)` and the requirement holds.  This is all to do with sorting by start position.